// ngx-image-compress/esm2015/lib/ngx-image-compress.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-image-compress.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2 } from '@angular/core';
import { ImageCompress } from './image-compress';
import { DOC_ORIENTATION } from './DOC_ORIENTATION';
import * as ɵngcc0 from '@angular/core';
export class NgxImageCompressService {
    /**
     * @param {?} rendererFactory
     */
    constructor(rendererFactory) {
        this.DOC_ORIENTATION = DOC_ORIENTATION;
        this.render = rendererFactory.createRenderer(null, null);
    }
    /**
     * @param {?} image
     * @return {?}
     */
    byteCount(image) {
        return ImageCompress.byteCount(image);
    }
    /**
     * Get the correct Orientation value from the EXIF tags in the specified file.
     * @param {?} file
     * @return {?}
     */
    getOrientation(file) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        (resolve) => {
            ImageCompress.getOrientation(file, (/**
             * @param {?} result
             * @return {?}
             */
            (result) => {
                resolve(result);
            }));
        }));
    }
    /**
     * @return {?}
     */
    uploadFile() {
        return ImageCompress.uploadFile(this.render);
    }
    /**
     * @param {?} image
     * @param {?} orientation
     * @param {?=} ratio
     * @param {?=} quality
     * @return {?}
     */
    compressFile(image, orientation, ratio = 50, quality = 50) {
        return ImageCompress.compress(image, orientation, this.render, ratio, quality);
    }
}
NgxImageCompressService.ɵfac = function NgxImageCompressService_Factory(t) { return new (t || NgxImageCompressService)(ɵngcc0.ɵɵinject(ɵngcc0.RendererFactory2)); };
NgxImageCompressService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgxImageCompressService, factory: NgxImageCompressService.ɵfac });
/** @nocollapse */
NgxImageCompressService.ctorParameters = () => [
    { type: RendererFactory2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxImageCompressService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.RendererFactory2 }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxImageCompressService.prototype.render;
    /** @type {?} */
    NgxImageCompressService.prototype.DOC_ORIENTATION;
}
